File: sdks/dart_client/lib/src/auth/session_store.dart
/// Pluggable persistence for the signed-in session.
///
/// The default ([InMemorySessionStore]) keeps the session for the life of the
/// process — fine for servers and tests. A Flutter app supplies a persistent
/// implementation (e.g. backed by `shared_preferences` or
/// `flutter_secure_storage`) so a user stays logged in across restarts. Keeping
/// this an interface means the SDK never forces a storage dependency on apps
/// that don't want one.
library;

/// Reads, writes, and clears the serialized session string.
abstract class SessionStore {
  Future<String?> read();
  Future<void> write(String value);
  Future<void> delete();
}

/// Non-persistent default. Lives only as long as the client.
class InMemorySessionStore implements SessionStore {
  String? _value;

  @override
  Future<String?> read() async => _value;

  @override
  Future<void> write(String value) async => _value = value;

  @override
  Future<void> delete() async => _value = null;
}

File: sdks/typescript_client/src/auth/session-store.ts
/**
 * Pluggable persistence for the signed-in session.
 *
 * The default ({@link InMemorySessionStore}) keeps the session for the life of
 * the process — fine for servers, route handlers, and tests. Web and React
 * Native apps supply a persistent store so a user stays signed in across
 * restarts; {@link persistentSessionStore} adapts both `localStorage` and
 * `AsyncStorage`. Keeping this an interface means the SDK never forces a
 * storage dependency on apps that don't want one.
 */

/** Reads, writes, and clears the serialized session string. May be sync or async. */
export interface SessionStore {
  read(): Promise<string | null> | string | null;
  write(value: string): Promise<void> | void;
  delete(): Promise<void> | void;
}

/** Non-persistent default. Lives only as long as the client. */
export class InMemorySessionStore implements SessionStore {
  private value: string | null = null;

  read(): string | null {
    return this.value;
  }

  write(value: string): void {
    this.value = value;
  }

  delete(): void {
    this.value = null;
  }
}

/**
 * The subset of `localStorage` / `@react-native-async-storage/async-storage`
 * this SDK uses. Both satisfy it (localStorage sync, AsyncStorage async).
 */
export interface KeyValueStorage {
  getItem(key: string): Promise<string | null> | string | null;
  setItem(key: string, value: string): Promise<void> | void;
  removeItem(key: string): Promise<void> | void;
}

/**
 * A persistent store backed by any {@link KeyValueStorage}.
 *
 * ```ts
 * // Next.js client component / web
 * persistentSessionStore(window.localStorage)
 * // React Native
 * import AsyncStorage from "@react-native-async-storage/async-storage";
 * persistentSessionStore(AsyncStorage)
 * ```
 */
export function persistentSessionStore(
  storage: KeyValueStorage,
  key = "forgefy.session",
): SessionStore {
  return {
    read: () => storage.getItem(key),
    write: (value) => storage.setItem(key, value),
    delete: () => storage.removeItem(key),
  };
}
